### ISSUE TITLE
New Feature: Rate limiting

### ISSUE THREAD

**Issue by jpillora** [anacrolix/torrent#11]
Are there plans to add rate limiting? alternatively, a connection hook might work, where one could implement rate limiting themselves. I don't know enough about the Bittorrent protocol though to know if this is feasible.

**Comment by anacrolix** [comment]
There are no current plans. I believe it's quite straightforward. It can be done via a connection hook. All writing and reading on the wire are done via `connection.rw`. That's already a series of hooks that take care of protocol encryption, TCP vs. uTP, as well as deadlines. You could add a hook either before any handshaking occurs (`client.{incoming,outgoing}Connection`), or during the message exchange loop (`client.connectionLoop`) as you see fit. Rate limiting on downloads would result in the peer doing the appropriate things with the protocol. Rate limiting on upload is a little harder. Currently `connection.writeOptimizer` will help significantly applying backpressure from inability to write due to rate limiting back to the client. There's probably some improvements that could be done there, but it's not necessary to get the basics working.

I tried to link to areas of the code via https://sourcegraph.com/github.com/anacrolix/torrent but it didn't work very well.

**Comment by jpillora** [comment]
Passing a facade `connection.rw` sounds like a good idea. Could `newConnection()` become `newConnection(nc net.Conn)`? Could then move `c.conn = nc` and `c.rw = nc` inside the constructor.

Then before you call `newConnection`:

``` go

type ConnectionHook func(nc net.Conn) net.Conn

func (cl *client) SetDownloadHook(hook ConnectionHook) {
}
func (cl *client) SetUploadHook(hook ConnectionHook) {
}

//outgoing
if cl.hookUpload != nil {
  nc = cl.HookUpload(nc)
}
c := newConnection(nc)


//incoming
if cl.hookDownload != nil {
  nc = cl.hookDownload(nc)
}
c := newConnection(nc)
```

Or hooks could just be public, and you set them directly.

I can try my hand at a PR if needed

**Comment by anacrolix** [comment]
I believe `newConnection` did take a bunch of arguments previously to completely instantiate itself in one go. I removed a lot of assumptions from it (like that it had to be connected immediately) so as to allow manipulating a lot of its optional state, and installing hooks as required. If you see in `receiveHandshakes` and `initiateHandshakes` that connection fields are then further manipulated, I thought it best not to give people the false assumption that fields like `rw` and `conn` were necessarily only assigned to in `newConnection`.

I think since connections can exist before torrents are even added, that perhaps the hooks could come in via the client Config. I see no reason to allow modifying the hooks once a client is already created via NewClient. This will let us simplify our management of the hooks. A PR would be great. When/where to install the hooks is up to you. That is whether you want them before or after handshakes, above or below the encryption/BitTorrent protocol or directly on top of the transport.

I'm also not sure if it's necessary to separate the ReadWriter into separate parts (ie by having Upload/Download separate). I'm not sure what's gained/lost from this. At least presently it's `rw io.ReadWriter`, but the read/write parts are used separately. Whatever works best.

**Comment by jpillora** [comment]
Hey @anacrolix sorry, haven't had time to look into this for a while. I've started my [torrent client](https://github.com/jpillora/cloud-torrent) and I've got a ways to go, though rate limiting still will be needed.

Check out the planned features https://github.com/jpillora/cloud-torrent and let me know if it's something you'd use or if you have any ideas - comment on the issues - we'll this issue just for rate limiting.

**Comment by anacrolix** [comment]
I have a similar private project. anacrolix/torrent is designed with supporting the features you require for that project in mind.

**Comment by jpillora** [comment]
Alrighty, well if you ever want to combine forces - let me know

**Comment by anacrolix** [comment]
@jpillora Note this was fixed in d4cbdc5c38ac835ceedc5e477bfc335f3e71652f and 836bb344f3602205275f9f6abbe9a986bb988e41.

**Comment by jpillora** [comment]
Ah cool 👍
